src/sctools/test/test_count.py

```diff
@@ -1,44 +1,578 @@
+        count_matrix: np.ndarray = self._generate_random_count_matrix()
+
+        # generate primary alignment tags that produce count_matrix
+        primary_alignment_record_tags_set, row_index, col_index = \
+            self._generate_primary_alignment_record_bundle(count_matrix)
+        primary_alignment_record_tags_list = list(primary_alignment_record_tags_set)
+
+        # sanity check -- we require as many primary alignment records as the total counts
+        assert len(primary_alignment_record_tags_set) == np.sum(count_matrix)
+
+        # add duplicate records
+        duplicate_alignment_tags_list = self._generate_duplicate_alignment_tags(
+            num_duplicates, primary_alignment_record_tags_list)
+
+        # add records with missing tags
+        incomplete_alignment_tags_list: List[AlignmentRecordTags] = self._generate_incomplete_alignment_tags(
+            num_missing_some_tags)
+
+        # add records with multiple gene alignments
+        multiple_alignment_tags_list: List[List[AlignmentRecordTags]] = self._generate_multiple_gene_alignment_tags(
+            num_multiple_gene_alignments,
+            max_gene_hits_per_multiple_gene_alignments,
+            primary_alignment_record_tags_set)
+
+        return SyntheticDataBundle(count_matrix,
+                                   row_index,
+                                   col_index,
+                                   primary_alignment_record_tags_list,
+                                   duplicate_alignment_tags_list,
+                                   incomplete_alignment_tags_list,
+                                   multiple_alignment_tags_list)
+
+    def generate_synthetic_bam_and_counts_matrix(self,
+                                                 output_path: str,
+                                                 num_duplicates: int,
+                                                 num_missing_some_tags: int,
+                                                 num_multiple_gene_alignments: int,
+                                                 max_gene_hits_per_multiple_gene_alignments: int,
+                                                 query_sort_order: str = 'CB_UB_GE_n'):
+        """Generates synthetic count matrix and BAM file and writes them to disk.
+
+        Parameters
+        ----------
+        output_path : str
+            output path
+        num_duplicates : int
+            number of duplicate records
+        num_missing_some_tags : int
+            number of records that miss at least one crucial tag
+        num_multiple_gene_alignments : int
+            number of records that have at least two different gene tags
+        max_gene_hits_per_multiple_gene_alignments : int
+            maximum number of unique gene names to use for multiple-gene records
+        query_sort_order : str
+            'unsorted' : shuffle the records before saving to BAM
+            'CB_UB_GE_n' : sort hierarchically by CB, UB, GE, and query_name
```

Comment:
What do you think about making this a `tuple`, and adding an `Enum` class somewhere that lists the valid sort fields in case we want the bam sorted in a different order? The gene metrics, for one, request a different sort order.

Reply:
I added `AlignmentSortOrder` base class and two instances to `bam.py`.